### PR TITLE
Use HTML5 sanitizer to serve huge payloads through Speedscope remote viewer

### LINKED ISSUE #92: Speedscope viewer fails for big payloads
:warning::rotating_light::warning: Project/labels may still need to be added :warning::rotating_light::warning:

## Summary

- David discovered a bug while profiling with AppProfiler, where the profile results could not be displayed in Spin due to an issue with large JSON files. This required the JSON file to be downloaded and opened locally with `speedscope`.
- He found that when the JSON file produced by AppProfiler is too large, the javascript tag containing the flamegraph’s data isn't completely rendered.
- David noticed that 75% of the string is lost after going through `Loofah.html4_fragment` and suspected that Nokogiri, which Loofah uses, may be breaking when a single line in the parsed HTML fragment is too long.
- Mike suggested setting the HUGE parse option on the HTML4 parser, or using the HTML5 parser which does not have soft limits. In a Rails app, if on Rails 7.1, the `config.action_view.sanitizer_vendor` can be set to `Rails::HTML5::Sanitizer` to use the HTML5 parser.
- David will check if this can be fixed in AppProfiler.

_This summary was generated using [OpenAI's gpt-4 with a temperature of 0.5](https://openai.com/blog/openai-api/)._

<details>
<summary>:thread: Slack Thread</summary>
<table>
  <thead>
    <tr>
      <th colspan="2" width="181px">User</th>
      <th width="619px">Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/davidstosik">David Stosik</a><br/>
      <time>2023&#8209;09&#8209;22&#160;10:06</time>
    </sup>
  </td>
  <td>
    I paired with @gmcgibbon yesterday to learn how to profile with AppProfiler, this was great, but I was nerd-sniped by the fact that we sometimes can’t show the profile results in Spin, so we need to download the JSON file and open it locally with <code>speedscope</code>. This is not a big deal but that shows there’s probably a bug somewhere. I dug a little bit and noticed that:
• AppProfiler’s “Speedscope remote viewer” produces the `` tag that contains the flamegraph’s data <a href="https://github.com/Shopify/app_profiler/blob/617ad9bc89a0e46c6b695f9eb86e589f0fa983a4/lib/app_profiler/viewer/speedscope_remote_viewer/middleware.rb#L36-L52">here</a>.
• When the JSON file produced by AppProfiler is too big, that javascript tag is not completely rendered. However, the served page source has closing ``  (<a href="https://github.com/Shopify/app_profiler/blob/617ad9bc89a0e46c6b695f9eb86e589f0fa983a4/lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb#L61-L62">generated here</a>) tags (so Rack is probably not the problem)
• the graph data read from the JSON file is passed through <code>Sanitizer#sanitize</code>, which is rails-html-sanitizers’ <code>Rails::Html::SafeListSanitizer#sanitize</code>  with a reduced list of allowed tags
•  I decomposed the <a href="https://github.com/rails/rails-html-sanitizer/blob/19fd6cd66f31316642e758bf01a410f3fd128f42/lib/rails/html/sanitizer.rb#L39">“composed sanitize” method</a> into each part to introduce logs (see in thread).
• This allowed me to notice we lose 75% of the string after going through <code>Loofah.html4_fragment</code>. Also, the resulting string’s size is suspiciously round (10,000,040 characters)
• I experienced with Nokogiri in isolation (which is what Loofah uses), and noticed that it seems to break, maybe when a single line in the parsed HTML fragment is too long? See in thread for the debug script.
@flavorjones Do you mind me picking your mind about Nokogiri? Are you aware of a line or data length limit which could break parsing?
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/davidstosik">David Stosik</a><br/>
      <time>2023&#8209;09&#8209;22&#160;10:07</time>
    </sup>
  </td>
  <td>
    My own logs from the “decomposed sanitize method”.

<hr>

<ul><li><a href="https://storage.googleapis.com/shopify-ihub-github-archiver/a68ff104714cefee8a6e8992dea5a98682cc836e9580c248de082f5f7b1773b1">Untitled</a></li>
</ul>
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/davidstosik">David Stosik</a><br/>
      <time>2023&#8209;09&#8209;22&#160;10:09</time>
    </sup>
  </td>
  <td>
    A minimal script to reproduce the Nokogiri problem, and its output.

<pre lang="">== RAW HTML ==
Length: 40000064
Occurrences of &#39;abc &#39;: 10000000
End of string:
c abc abc abc abc abc abc abc abc abc &quot;

== PARSED HTML ==
Length: 10000040
Occurrences of &#39;abc &#39;: 2499994
End of string:
c abc abc abc abc abc abc abc abc abc abc</pre>

<hr>

<ul><li><a href="https://storage.googleapis.com/shopify-ihub-github-archiver/f596488ee217d0ab4b97440f476e81168ddb6341529948c3ae15b18d16eeebe5.txt">long-line_rb.txt</a></li>
</ul>
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/flavorjones">Mike Dalessio</a><br/>
      <time>2023&#8209;09&#8209;22&#160;12:39</time>
    </sup>
  </td>
  <td>
    probably have to <a href="https://nokogiri.org/rdoc/Nokogiri/XML/ParseOptions">set the HUGE parse option</a> on the HTML4 parser
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/flavorjones">Mike Dalessio</a><br/>
      <time>2023&#8209;09&#8209;22&#160;12:39</time>
    </sup>
  </td>
  <td>
    or use the HTML5 parser which does not have soft limits
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/flavorjones">Mike Dalessio</a><br/>
      <time>2023&#8209;09&#8209;22&#160;12:43</time>
    </sup>
  </td>
  <td>
    in a rails app, if you're on rails 7.1, set <code>config.action_view.sanitizer_vendor</code> to <code>Rails::HTML5::Sanitizer</code> to use the HTML5 parser (see <a href="https://edgeguides.rubyonrails.org/configuring.html#config-action-view-sanitizer-vendor">https://edgeguides.rubyonrails.[…]sanitizer-vendor</a>)
  </td>
</tr>
<tr>
  <th></th>
  <td>
    <sup>
      <a href="https://github.com/davidstosik">David Stosik</a><br/>
      <time>2023&#8209;09&#8209;25&#160;07:45</time>
    </sup>
  </td>
  <td>
    Oh that sounds right! Thank you very much! Will check if we can fix that in AppProfiler.
  </td>
</tr>
  </tbody>
</table>
</details>

<hr>

<a href="https://github.com/davidstosik">David Stosik</a> archived this <a href="https://shopify.slack.com/archives/CM5HXA9AR/p1695377203727389?thread_ts=1695377203.727389&amp;cid=CM5HXA9AR">conversation from <code>team-ruby-and-rails-infra</code></a> at <time>2023&#8209;09&#8209;25&#160;08:11</time>.
All times are in UTC.

### PR DESCRIPTION
**Pull request by davidstosik**
Fixes #92.

When we run the app profiler on Shopify core's boot, it is frequent that the generated profile JSON file exceeds 10MB.

`Rails::Html::Sanitizer` is unable to process such huge payloads (because libxml2 limits text nodes to 10MB by default?).

The `rails-html-sanitizer` gem introduced HTML5 parsing in its 1.6.0 version, which lifts the text node size limit.

This change intends to use the new HTML5 parser when it's available to sanitize the Speedscope remote viewer's HTML, allowing bigger payloads to be served successfully.

I wasn't sure how to implement such variable superclass (depending on the version of rails-html-sanitizer used), and I was even less sure of how I could implement some tests to avoid regressions. All suggestions will be appreciated! 🙏🏻

### CODE REVIEW

File: test/app_profiler/viewer/speedscope_remote_viewer/middleware_test.rb [hunk 1]
```diff
@@ -60,6 +60,21 @@ class MiddlewareTest < TestCase
           assert_match(%r(<script type="text/javascript">), html)
         end
 
+        test "#call show can serve huge payloads" do
+          frames = { "1" => { name: "a" * 1e7 } }
```
Comment (davidstosik) [inline test/app_profiler/viewer/speedscope_remote_viewer/middleware_test.rb:64]: 10MB, or about 10 million (`1e7`) characters, is the size limit the old HTML4 sanitizer wasn't able to exceed.
Reverting `rails-html-sanitizer` to 1.5.0 in Gemfile.lock, one can run this test and confirm that it fails.

File: lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb [hunk 2]
```diff
@@ -6,7 +6,13 @@ module AppProfiler
   module Viewer
     class SpeedscopeRemoteViewer < BaseViewer
       class BaseMiddleware
-        class Sanitizer < Rails::Html::SafeListSanitizer
+        sanitizer_superclass = if Rails::Html::Sanitizer::VERSION >= "1.6.0" && Rails::HTML::Sanitizer.html5_support?
```
[the remark this answers is not in the document]
Comment (flavorjones) [reply lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb:9]: You can/should use `Rails::HTML::Sanitizer.best_supported_vendor` for this if you pin to RHS >= 1.6.0

File: Gemfile.lock [hunk 3]
```diff
@@ -104,35 +104,36 @@ GEM
...
+    racc (1.7.1)
     rack (2.2.6.3)
     rack-test (1.1.0)
       rack (>= 1.0, < 3)
     rails-dom-testing (2.0.3)
       activesupport (>= 4.2.0)
       nokogiri (>= 1.6)
-    rails-html-sanitizer (1.5.0)
-      loofah (~> 2.19, >= 2.19.1)
+    rails-html-sanitizer (1.6.0)
```
[the remark this answers is not in the document]
Comment (flavorjones) [reply Gemfile.lock:134]: Why is it important to support RHS 1.5.x? I'd prefer if you just updated the dependency.

File: lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb [hunk 4]
```diff
@@ -6,7 +6,13 @@ module AppProfiler
...
+          Rails::HTML5::SafeListSanitizer
+        else
+          Rails::Html::SafeListSanitizer
```
[the remark this answers is not in the document]
Comment (flavorjones) [reply lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb:12]: See previous comments on my recommendation to use 1.6.0

**Review by flavorjones — COMMENTED** [review]
I want to gently push on the assumption that we need to support Rails::HTML::Sanitizer v1.5.x. v1.6.0 should be backwards-compatible and supports Ruby 2.7 and later.

**Comment by davidstosik** [comment]
@flavorjones I'm unsure, but it looks like `rails-html-sanitizer` was not declared as a direct dependency of the `app_profiler` Gem, so I made the assumption that the gem was expected to work in any Rails application it's dropped in (including older Rails apps that might still be using `rails-html-sanitizer` 1.5.x).

I also guess we might not want to declare `rails-html-sanitizer` as a dependency in the gemspec because this gem could totally profile an app and produce a profile JSON file without it?

If we're happy to stop supporting older versions of `rails-html-sanitizer`, then I'll be happy to simplify the code, though I'm unsure how we'd gracefully ask users to update `rails-html-sanitizer`. 🤔

**Comment by flavorjones** [comment]
@davidstosik I understand your points, and you're right about the historical expectation. But what I'm suggesting is that we don't care anymore about supporting old versions of RHS (because the new version is backwards compatible) and that the right thing to do is to declare a dependency on RHS >= 1.6.0 and simplify this code.

File: lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb [hunk 5]
```diff
@@ -1,12 +1,13 @@
 # frozen_string_literal: true
 
+gem "rails-html-sanitizer", ">= 1.6.0"
```
Comment (davidstosik) [inline lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb:3]: @flavorjones To enforce the version dependency, I had to add a peer dependency here. (There is no dependency to that gem in the gemspec.)

**Review by gmcgibbon — APPROVED** [review]

**Comment by davidstosik** [comment]
@shioyama: I paired with @gmcgibbon this morning and we confirmed that this change also fixes the double escaping issue you mentioned to me yesterday! 🎉 

| Before | After |
|---|---|
| image | image |

**Pull request merged**